### ISSUE TITLE
ICO's cotribute/refund/claim algorithm proposal

### ISSUE THREAD

**Issue by Kiwonik** [steemit/steem#2396]
This issue contains proposal that implements and impacts SMT ICO contribution/refund/claim operations/actions, therefore separated from specific issues of these entities.

The core ideas are:
1. Selected data of source contribution operations is stored in container associated with given SMT object (the `source_contribution_container`).
   * This container is FIFO iterated only on successful ICO.
1. Another container stores cumulative contribution/distribution data associated per user (the `contributor_info_container`).
   * This container is iterated at ICO conclusion whether successful or not.
1. On ICO conclusion one automatic action replaces separate claim/refund actions.

The rationale and algorithm description:

1. Keep bare minimum source contribution data, copying only account name and amount from `smt_contribute_operation` to ordered `source_contribution_container`. 
This will allow for fast processing of source contributions in FIFO order without accessing the blocks where contribution (and refund) operations occurred.
1. Using `contributor_info_container` cumulative contribution numbers per contributor will greatly reduce the time of manual refund and automatic refund/claim processing (see below).
1. Use separate `total_contribution_number`, keeping track of SMT's cumulative contribution number for the same performance reasons.
1. When `smt_contribute_operation` is evaluated successfully, add a record to `source_contribution_container` and update contributor's total amount in `contributor_info_container`. Increase SMT's `total_contribution_number` accordingly.
1. In time window when manual `smt_refund_operation` is allowed (between announced launch time and expiration launch time, with some cap not revealed) after successful evaluation decrease both contributor's and SMT's total contribution numbers accordingly.
1. In time window when ICO may succeed (announced launch time passes and caps are revealed) checking whether enough contributions were made to surpass min cap is as easy as comparing the cap to `total_contribution_number`. When comparison is positive, move to FIFO processing of source contribution data in a loop over contributions:
   1. Process `min(contribution.amount, user.total_contribution)` deducting the amount from user's total contribution and increasing appropriate user's awarded token amount (according to distribution policy). Note that cumulative distrubution numbers are stored in `contributor_info_container` too.
      * If the amount surpasses one of the caps, do processing in two separate takes - for amount fraction below and above the cap.
   1. Break loop when surpassed hard cap.
   1. Distribute SMTs and refund remaining contributions (if any) in single iteration over `contributor_info_container`, thus executing automatic refund/claim action in one take.
1. When ICO fails, do the same as above ^ i.e. iterate once over `contributor_info_container` doing automatic refunds.
1. Automatic actions scheduling/execution in tranches/batches is easily applied on the algorithm to conform with resource usage policy.

@mvandeberg @theoreticalbts @youkaicountry

**Comment by mvandeberg** [comment]
Let me summarize the data structures involved to ensure that I understand the proposal.

We have 3 structs and 2 data structures.

1. The `total_contribution_number` contained in an SMT specific ICO struct.
2. `smt_contribution` corresponds `1:1` with `smt_contribution_operation` and is stored FIFO.
3. `contributor_info` corresponds `1:n` with `smt_contribution_operation` and is an aggregate of a user's contributions.

The advantage that this algorithm gives us, which I don't see explicitly stated, is that is allows for a contributor to refund the entirety of their contributions with a single op. However, this isn't actually more efficient. We can short cut the initial validation using the `contributor_info`, but we still have to remove all `n` contributions.

Let's say Alice makes two contributions to an ICO with hidden soft and hard caps. One contribution is below the soft cap and the other is between the soft and hard caps. Let's also say Bob made a contribution above the hard cap and the removal of Alice's contribution would place him below the hard cap. Charlie also made a contribution between the soft and hard caps and his contribution would be below the soft cap if Alice were to ask for a refund. If the launch time passes without the caps being revealed, Alice can choose to be refunded. If she does then the initial tokens both Bob and Charlie receive change. The only way to know this is to iterate over the `smt_contribution_container`.

When Alice asks for a refund we can iterate over the `smt_contribution_container`, removing contributions and validating her total, without needing the `contibutor_info`. However, if we have the `contributor_info`, we can refund Alice instantly at the cost of having to validate each contribution at the end of the ICO.

I do have one question regarding the intended behavior described in the white paper. Is the refund operation intended as an all or nothing refund? Perhaps Alice only wants to refund her second contribution, should she be allowed to? In such a scenario, this would increase the validation logic of each `smt_contribution` as we have to track the totals against the `contributor_info` to know when the sum of the contributions pass the actual total contributed. In which case, I think it makes more sense to update the `smt_contributions` as we go. Adding another index that is FIFO on the `[smt, account]` tuple would significantly simplify this logic.

My conclusion is that at this time we need to answer my question regarding the intended behavior of the refund operation. The rest of this looks like premature optimization to me at this point as each implementation is the same order of complexity.

**Comment by Kiwonik** [comment]
@mvandeberg My clarification step by step:
> The advantage that this algorithm gives us, which I don't see explicitly stated, is that is allows for a contributor to refund the entirety of their contributions with a single op. However, this isn't actually more efficient. We can short cut the initial validation using the `contributor_info`, but we still have to remove all `n` contributions.

Yes, a contributor can refund entirety or any part of their contributions no matter the number or sizes of their contribute operations. No, we don't remove any contributions at this refund, no change to `smt_contribution_container` is made,  only a single lookup in `contributor_info_container` to decrease the contributor's total amount there. This is where performance gain is made. Also, see below.

> Let's say Alice makes two contributions to an ICO with hidden soft and hard caps. One contribution is below the soft cap and the other is between the soft and hard caps. Let's also say Bob made a contribution above the hard cap and the removal of Alice's contribution would place him below the hard cap. Charlie also made a contribution between the soft and hard caps and his contribution would be below the soft cap if Alice were to ask for a refund. If the launch time passes without the caps being revealed, Alice can choose to be refunded. If she does then the initial tokens both Bob and Charlie receive change. The only way to know this is to iterate over the smt_contribution_container.
> 
> When Alice asks for a refund we can iterate over the smt_contribution_container, removing contributions and validating her total, without needing the contibutor_info. However, if we have the contributor_info, we can refund Alice instantly at the cost of having to validate each contribution at the end of the ICO.

Without `contributor_info` we have to iterate over `smt_contribution_container` every time a contributor requests a refund. Then we have to iterate over it once more at ICO success or failure to do automatic claim or refund.

With `contributor_info` there's no iteration at refund request (operation). At successful ICO we do iterate over `smt_contribution_container`, but at ICO failure we iterate over `contributor_info_container` of which size is less or equal the size of `smt_contribution_container`.

In other words, the more refund operations contributors make, the more we gain using `contributor_info`.

**Comment by mvandeberg** [comment]
I would still like clarification as to the behavior of a refund in the white paper. It is implied at places that the refund can be partial and others that it is not, but never explicitly stated. @youkaicountry @theoreticalbts 

Another solution, as I stated earlier is adding another index on `smt_contribution` for `[smt, account]`. We could get rid of `contibutor_info` entirely. Yes, each refund would require iterating through the smt contributions for that particular contributor, but it also allows us to remove the objects and not have to iterate over them later. Let's say this is O(c) for the number of contributions an account makes.

Then, when triggering the initial emissions it is an O(n - c) iteration. And we clear up memory in the meantime.

Also consider the scenario when an ICO never triggers the initial emission and all contributors get a refund. All of their contribution objects are in the database and will never be removed.

You can make the argument that we don't want to add another index on the contribution objects, but in my proposal I removed the contributors object (and index on such object) entirely.

**Comment by Kiwonik** [comment]
@mvandeberg As both proposals presented here employ means of reducing complexity of manual refund evaluation (`contributor_info` or additional index on `smt_contribution`) I proposed in #2054 that the `contribution_id` fields are removed from contribute, ~claim~ and refund operations as I understand they were meant for the same purpose. However @theoreticalbts has some doubts about it, so I suggested here is the better place to make sure they are no longer needed. Do you agree they are no longer needed no matter which algorithm is accepted here?

**Comment by vogel76** [comment]
@mvandeberg @theoreticalbts 
1. Our intention to create another data structure `contributor_info_container` was to significantly reduce complexity of manual refund operation, which in your solution has (at least) linear complexity and can lead to DOS attack (this argument has been many times opened during automatic action discussion).  Also removing contribution infos objects during `smt_refund_operation` evaluation increases time needed to handle them. Our solution's complexity does not depend on contribution number.
1. Additional memory footprint needed for `contributor_info_container` is not significant and can have small negative impact when every contributor made one contribution. Defining another index in the original `smt_contribution` container can have negative impact, when malicious contributor will produce a lot of tiny contributions (then holding additional tree structure for each contribution will use more memory than single entry in separate container).
1. State representation (`smt_contribution`,  `contributor_info_container`) shall be removed after ICO completes (does not matter if it succeeds or fails). In my opinion, every part (contributor and SMT-creator) shall have access to the list of contributions they were involved to, but this functionality can be covered by regular account-history features just by listing `smt_contribute_operations`. And this way all occupied memory (persistent storage) for the state representation of contributions will be cleaned up.

**Comment by youkaicountry** [comment]
Outdated.
